scene/3d/collision_object_3d.cpp

```diff
@@ -171,30 +183,34 @@ void CollisionObject3D::_update_pickable() {
 	}
 }
 
+bool CollisionObject3D::_are_collision_shapes_visible() {
+	return is_inside_tree() && get_tree()->is_debugging_collisions_hint() && !Engine::get_singleton()->is_editor_hint();
+}
+
 void CollisionObject3D::_update_debug_shapes() {
 	for (Set<uint32_t>::Element *shapedata_idx = debug_shapes_to_update.front(); shapedata_idx; shapedata_idx = shapedata_idx->next()) {
 		if (shapes.has(shapedata_idx->get())) {
 			ShapeData &shapedata = shapes[shapedata_idx->get()];
 			ShapeData::ShapeBase *shapes = shapedata.shapes.ptrw();
 			for (int i = 0; i < shapedata.shapes.size(); i++) {
 				ShapeData::ShapeBase &s = shapes[i];
-				if (s.debug_shape) {
-					s.debug_shape->queue_delete();
-					s.debug_shape = nullptr;
-					--debug_shape_count;
-				}
 				if (s.shape.is_null() || shapedata.disabled) {
+					if (s.debug_shape.is_valid()) {
+						RS::get_singleton()->free(s.debug_shape);
+						s.debug_shape = RID();
+						--debug_shapes_count;
+					}
 					continue;
 				}
-
+				if (s.debug_shape.is_null()) {
+					s.debug_shape = RS::get_singleton()->instance_create();
+					RS::get_singleton()->instance_set_scenario(s.debug_shape, get_world_3d()->get_scenario());
+					++debug_shapes_count;
+				}
 				Ref<Mesh> mesh = s.shape->get_debug_mesh();
-				MeshInstance3D *mi = memnew(MeshInstance3D);
-				mi->set_transform(shapedata.xform);
-				mi->set_mesh(mesh);
-				add_child(mi);
-				mi->force_update_transform();
-				s.debug_shape = mi;
-				++debug_shape_count;
+				s.shape->connect("changed", callable_mp(this, &CollisionObject3D::_update_shape_data), varray(shapedata_idx->get()), CONNECT_ONESHOT | CONNECT_REFERENCE_COUNTED);
```

Comment:
This function may be called multiple times, I think here you should check if the connection already exists, otherwise it will spam some error messages. If you are sure this never happens because of some logic, please add a comment line that explains how it works, this will turn useful in future.

Reply:
I have not tested every possible scenario, but the issue with that connect is when the same shape is added several times (I know it's an edge case, but I don't see any reason of why it shouldn't be handled). So, I decide to do the following:

- As before, I accumulated shape owners so _update_debug_shapes is called few times as possible.
- I added the CONNECT_ONESHOT flag, so when the signal is emitted it will be automatically disconnected, it will be reconnected at next _update_debug_shapes anyway.
- I added the CONNECT_REFERENCE_COUNTED flag, so the same shape can be used in different owners.
- When the ShapeData is removed, all the connections will be cleared.

Reply:
It seems like it would be much easier and safer to just check if the signal is connected. Are you avoiding the check for performance reasons? Since it's debug code, the extra gain doesn't matter much.

Reply:
After doing some tests with that particular scenario, I see that it is not working, so removing. My bad to not test it deeply.

Reply:
> Still, when the same shape is added to different owners, and the shape changes, only one occurrence of it will show.

In this case, this code could instead connect to a specific method that takes the shape itself as an argument and updates any debug shape that uses this shape.

Reply:
Ok, now it works as I want.